### PR TITLE
Add regression tests for regenerator

### LINKED ISSUE #6923: TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)
Choose one: bug report

The error message

> TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)

is shown when transpiling the input code. node successfully executes the input code.

### Input Code

```js
async function foo() {
	(async function (number) {
		const tmp = number
	})
}
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
{ "presets": [ [ "@babel/env"
               , { "targets": { "browsers": [ "last 2 chrome versions"
                                            , "last 2 chromeandroid versions"
                                            , "firefox esr"
                                            , "last 2 firefox versions"
                                            , "edge >= 15"
                                            , "safari >= 11"
                                            ]
                              }
                 , "debug": true
                 }
               ]
             ]
}
```

```js
{
  "dependencies": {
    "@babel/cli": "^7.0.0-beta.32",
    "@babel/core": "^7.0.0-beta.32",
    "@babel/preset-env": "^7.0.0-beta.32",
    "regenerator-runtime": "^0.11.0"
  }
}
```

### Expected Behavior

The script is successfully transpiled.

### Current Behavior

The given error message appears:

```
root@3f46f465fd04:/pwd# yarn run babel test.js
yarn run v1.3.2
warning package.json: No license field
$ /pwd/node_modules/.bin/babel test.js
@babel/preset-env: `DEBUG` option

Using targets:
{
  "chrome": "61",
  "edge": "15",
  "firefox": "52",
  "safari": "11"
}

Using modules transform: commonjs

Using plugins:
  transform-destructuring { "edge":"15", "firefox":"52" }
  transform-for-of { "firefox":"52" }
  transform-function-name { "edge":"15", "firefox":"52" }
  transform-literals { "firefox":"52" }
  transform-parameters { "firefox":"52" }
  transform-regenerator { "firefox":"52" }

Using polyfills: No polyfills were added, since the `useBuiltIns` option was not set.
TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)
    at File.buildCodeFrameError (/pwd/node_modules/@babel/core/lib/transformation/file/file.js:209:12)
    at Scope.checkBlockScopedCollisions (/pwd/node_modules/@babel/traverse/lib/scope/index.js:295:27)
    at Scope.registerBinding (/pwd/node_modules/@babel/traverse/lib/scope/index.js:477:16)
    at Scope.registerDeclaration (/pwd/node_modules/@babel/traverse/lib/scope/index.js:396:14)
    at Object.Declaration (/pwd/node_modules/@babel/traverse/lib/scope/index.js:89:12)
    at NodePath._call (/pwd/node_modules/@babel/traverse/lib/path/context.js:64:19)
    at NodePath.call (/pwd/node_modules/@babel/traverse/lib/path/context.js:38:17)
    at NodePath.visit (/pwd/node_modules/@babel/traverse/lib/path/context.js:95:12)
    at TraversalContext.visitQueue (/pwd/node_modules/@babel/traverse/lib/context.js:139:18)
    at TraversalContext.visitMultiple (/pwd/node_modules/@babel/traverse/lib/context.js:93:17)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Context

It worked fine with `babel-preset-env` in version `1.5.2`. It is broken with `@babel/preset-env` in version `7.0.0-beta.32`.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.32 (@babel/core 7.0.0-beta.32)
| Babylon          | 
| node             | v9.2.0
| npm              | 5.5.1
| Operating System | Official node.js Docker container

### LINKED ISSUE #8525: `Duplicate declaration` error with Babel 7, React preset and regenerator transform plugin.
## Bug Report

**Current Behavior**
```
$ babel ./test.js --no-babelrc --presets=module:@babel/preset-react --plugins=module:@babel/plugin-transform-regenerator
TypeError: /Users/stevefar/sc/labeltron/test.js: Duplicate declaration "newSessionId" (This is an error on an internal node. Probably an internal error.)
    at File.buildCodeFrameError (/Users/stevefar/sc/labeltron/node_modules/@babel/core/lib/transformation/file/file.js:259:12)
    at Scope.checkBlockScopedCollisions (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/scope/index.js:347:22)
    at Scope.registerBinding (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/scope/index.js:504:16)
    at Scope.registerDeclaration (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/scope/index.js:444:14)
    at Object.Declaration (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/scope/index.js:125:12)
    at NodePath._call (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/path/context.js:53:20)
    at NodePath.call (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/path/context.js:40:17)
    at NodePath.visit (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/path/context.js:88:12)
    at TraversalContext.visitQueue (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/context.js:118:16)
    at TraversalContext.visitMultiple (/Users/stevefar/sc/labeltron/node_modules/@babel/traverse/lib/context.js:85:17)
```

**Input Code**
The is the contents of test.js

```js
<Component>
  {() => {
    const saveSession = async () => {
      const newSessionId = await someAsyncFunc();
    };
  }}
</Component>;

```

**Expected behavior/code**
I get a runtime error running `babel`, I expected it to create the transpiled code.

**Babel Configuration (.babelrc, package.json, cli command)**

babel ./test.js --no-babelrc --presets=module:@babel/preset-react --plugins=module:@babel/plugin-transform-regenerator

**Environment**
- Babel version(s): 7.0.0-rc.2 (@babel/core 7.0.0-rc.2)
- Node/npm version: Node 8/npm 5
- OS: OSX 10.13.6
- Monorepo no
- How you are using Babel: cli

**Additional context/Screenshots**
If I do each transform as a individual step it works:

```
$ cat ./test.js | babel --no-babelrc --presets=module:@babel/preset-react | babel --no-babelrc --plugins=module:@babel/plugin-transform-regenerator

```

Its not clear why this would work, but including both the preset and plugin would not.

### LINKED ISSUE #10193: Variable assignment while object destructuring not working properly 'Error: "action" is read-only.'
## Bug Report

**Current Behavior**
We have a promise rejection in our react-native app.
It is possible to log it thru logcat (android logs).
It is regarding following code: See Input Code

When I try to "overload" the 'someAction' variable I get an runtime error which I can inspect via logcat (android logs used on my react-native app) looking something like this:

`
Possible Unhandled Promise Rejection (id: 1):
    Error: "someAction" is read-only.
`

If I don't "overload" the variable and instead try to do something like this
`
  componentWillMount() {
    this.props.someAction();
  }
`
it will work.

I guess it has something to do with how the babel helper handles the arguments and/or the export/imports.
But I am not sure at all.

**Input Code**

```js
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { someAction } from '../../store/data/actions';

class SomeComponent extends Component {
  componentWillMount() {
    this.foo();
  }

  async foo() {
    const { someAction } = this.props;
    return await someAction();
  }

  render() {
    return null;
  }
}

export default connect(null, {
  someAction,
})(SomeComponent);
```

**Expected behavior/code**
Expected behavior would be that I can use a variable name inside my class/component (i.e. getting my action out of my props attribute by destructuring), even tho the variable name is also taken by the import of the action in the file. 

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "presets": ["module:metro-react-native-babel-preset"],
  "env": {
    "production": {
      "plugins": ["transform-remove-console"]
    }
  }
}

    "@babel/runtime": "^7.5.4",
    "babel-plugin-transform-remove-console": "^6.9.4",
    "metro-react-native-babel-preset": "^0.55.0",
```

**Environment**
- Babel version(s): 7.5.4
- Node/npm version: Node v10.15.3 / Npm 6.4.1
- OS: OSX 10.14.5
- How you are using Babel: `react-native`

**Additional context/Screenshots**
Babel transpiles the code to something like this:
```js
__d(function(global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
    var _interopRequireDefault = _$$_REQUIRE(_dependencyMap[0], "@babel/runtime/helpers/interopRequireDefault");
    var _interopRequireWildcard = _$$_REQUIRE(_dependencyMap[1], "@babel/runtime/helpers/interopRequireWildcard");
    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.default = void 0;
    var _regenerator = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[2], "@babel/runtime/regenerator"));
    var _classCallCheck2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[3], "@babel/runtime/helpers/classCallCheck"));
    var _createClass2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[4], "@babel/runtime/helpers/createClass"));
    var _possibleConstructorReturn2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[5], "@babel/runtime/helpers/possibleConstructorReturn"));
    var _getPrototypeOf2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[6], "@babel/runtime/helpers/getPrototypeOf"));
    var _inherits2 = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[7], "@babel/runtime/helpers/inherits"));
    var _react = _interopRequireWildcard(_$$_REQUIRE(_dependencyMap[8], "react"));
    var _index = _interopRequireDefault(_$$_REQUIRE(_dependencyMap[9], "./../../../node_modules/react-transform-hmr/lib/index.js"));
    var _reactRedux = _$$_REQUIRE(_dependencyMap[10], "react-redux");
    var _actions = _$$_REQUIRE(_dependencyMap[12], "../../store/data/actions");
    var _services = _$$_REQUIRE(_dependencyMap[15], "../../services");
    var _jsxFileName = "/somePath/node_modules/react-native/app/components/SomeComponent/SomeComponent.js";
    var _components = {
        SomeComponent: {
            displayName: "SomeComponent"
        }
    };
    var _node_modulesReactTransformHmrLibIndexJs2 = (0, _index.default)({
        filename: "/somePath/node_modules/react-native/app/components/SomeComponent/SomeComponent.js",
        components: _components,
        locals: [module],
        imports: [_react.default]
    });

    function _wrapComponent(id) {
        return function(Component) {
            return _node_modulesReactTransformHmrLibIndexJs2(Component, id);
        };
    }
    var SomeComponent = function(_Component) {
        (0, _inherits2.default)(SomeComponent, _Component);

        function SomeComponent() {
            (0, _classCallCheck2.default)(this, SomeComponent);
            return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(SomeComponent).apply(this, arguments));
        }(0, _createClass2.default)(SomeComponent, [{
            key: "componentWillMount",
            value: function componentWillMount() {
                this.foo();
            }
        }, {
            key: "foo",
            value: function foo() {
                var _this$props, someAction, _ref, type;
                return _regenerator.default.async(function foo$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                _this$props = this.props, _actions.someAction = (_this$props.someAction, function() {
                                    throw new Error('"' + "someAction" + '" is read-only.');
                                }());
                                return _regenerator.default.awrap((0, _actions.someAction)());
                            case "end":
                                return _context.stop();
                        }
                    }
                }, null, this);
            }
        }, {
            key: "render",
            value: function render() {
                return null;
            }
        }]);
        return SomeComponent;
    }(_react.Component);
    SomeComponent = _wrapComponent("SomeComponent")(SomeComponent);
    var _default2 = (0, _reactRedux.connect)(null, {
        someAction: _actions.someAction,
    })(SomeComponent);
    exports.default = _default2;
}, 1365, [1, 544, 208, 114, 115, 122, 123, 125, 137, 255, 539, 575, 581, 956, 543, 584], "app/components/SomeComponent/SomeComponent.js");
```

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #10193, closes #6923, closes #8525
| Tests Added + Pass?      | Yes
| Any Dependency Changes?  | `regenerator` updated
| License                  | MIT

### CODE REVIEW

File: packages/babel-plugin-transform-regenerator/test/fixtures/scope/10193/output.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+/*#__PURE__*/
+regeneratorRuntime.mark(foo);
+
+function foo() {
+  var someAction;
+  return regeneratorRuntime.wrap(function foo$(_context) {
+    while (1) switch (_context.prev = _context.next) {
+      case 0:
+        _actions.someAction = (bar, function () {
+          throw new Error('"' + "someAction" + '" is read-only.');
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-regenerator/test/fixtures/scope/10193/output.js]: This is wrong, waiting for https://github.com/facebook/regenerator/pull/377

File: packages/babel-plugin-transform-regenerator/package.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
   "repository": "https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-regenerator",
   "main": "lib/index.js",
   "dependencies": {
-    "regenerator-transform": "^0.14.0"
+    "regenerator-transform": "^0.14.2"
```
Comment (thiagoarrais) [inline packages/babel-plugin-transform-regenerator/package.json]: Seems like https://github.com/facebook/regenerator/pull/377 is available in `regenerator-transform:0.14.1` shouldn't this be 0.14.1?

```suggestion
    "regenerator-transform": "^0.14.1"
```

Comment (nicolo-ribaudo) [reply packages/babel-plugin-transform-regenerator/package.json]: It isn't: https://github.com/facebook/regenerator/commits/master
- facebook/regenerator@ffc74ccb098e30c28c374fd114ce9ea9cb5614fe
- facebook/regenerator@d4ccc3ba278753c7c37e2db2147834007b9075b1

Comment (thiagoarrais) [reply packages/babel-plugin-transform-regenerator/package.json]: You're right. Sorry. My bad.

**Comment by nicolo-ribaudo** [comment]
@benjamn Could a new version of `regenerator-transform` including https://github.com/facebook/regenerator/commit/ffc74ccb098e30c28c374fd114ce9ea9cb5614fe be released? :pray:

**Comment by nicolo-ribaudo** [comment]
Regenerator has been merged in https://github.com/babel/babel/pull/11191, I'll merge these tests.

**Pull request merged**